### PR TITLE
[#11071] Return HasResponses for all sessions in a course to reduce QPS

### LINKED ISSUE #11071: Burst of concurrent requests causing high amount of instance uptime
**Environment**: Production
 
**Description**
The production server has been hit by high amount of instance uptime in recent days because a course with 50 open sessions is now ongoing. The guilty page is student home page: it will fetch the response status of all sessions in the table in parallel. 50 parallel requests is an almost 100% guarantee to spin up new instance.

**Proposed solutions**:
- Batching: Group individual request into batches of reasonable size and fire the API calls with the batches
- Sequentialization: Make the calls synchronous

### PR DESCRIPTION
**Pull request by moziliar**
Fixes #11071 

**Outline of Solution**

Allow session name to be null when the frontend intends to request for all sessions in a course. Adjusted `HasResponsesData` to include a map of all sessions and their respective `hasResponses` attribute.

Considering deprecating `hasStudentResponseForFeedbackSession` function has it is probably redundant to keep it; using the new `hasStudentResponseForAllFeedbackSessionsInCourse` would fetch all information needed for frontend at the moment and filtering can be done to extract the info of any sessions needed.

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/GetHasResponsesAction.java [hunk 1]
```diff
@@ -55,11 +59,21 @@ void checkSpecificAccessControl() {
...
+            String feedbackSessionName = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+            if (feedbackSessionName == null) {
+                List<FeedbackSessionAttributes> feedbackSessions = logic.getFeedbackSessionsForCourse(courseId);
+                if (feedbackSessions.isEmpty()) {
+                    // Course has no sessions and therefore no response; access to responses is safe for all.
+                    return;
+                }
+                gateKeeper.verifyAccessible(
+                        logic.getStudentForGoogleId(courseId, userInfo.getId()),
+                        feedbackSessions.get(0));
```
Comment (daongochieu2810) [inline src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:71]: why do we only check access for the first session?

File: src/main/java/teammates/ui/webapi/GetHasResponsesAction.java [hunk 2]
```diff
@@ -87,7 +101,20 @@ JsonResult execute() {
...
-            String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+            String feedbackSessionName = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+            if (feedbackSessionName == null) {
+                // check all sessions in the course
+                List<FeedbackSessionAttributes> feedbackSessions = logic.getFeedbackSessionsForCourse(courseId);
+                StudentAttributes student = logic.getStudentForGoogleId(courseId, userInfo.getId());
+
+                Map<String, Boolean> sessionsHasResponses = new HashMap<>();
+                for (FeedbackSessionAttributes feedbackSession : feedbackSessions) {
+                    boolean hasResponses = logic.hasStudentSubmittedFeedback(feedbackSession, student.email);
```
Comment (daongochieu2810) [inline src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:112]: maybe we can inline this value inside the method below

**Review by daongochieu2810 — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/ui/webapi/GetHasResponsesAction.java (hunk 1 above)
Comment (moziliar) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:71]: Good catch; I suppose we need to verify for all sessions indeed.

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts [hunk 3]
```diff
@@ -537,7 +537,7 @@ export class InstructorSessionEditPageComponent extends InstructorSessionBasePag
   private loadResponseStatusForQuestion(model: QuestionEditFormModel): void {
     this.feedbackSessionsService.hasResponsesForQuestion(model.feedbackQuestionId)
         .subscribe((resp: HasResponses) => {
-          model.isQuestionHasResponses = resp.hasResponses;
+          model.isQuestionHasResponses = resp.hasResponses[''];
```
Comment (teikjun) [inline src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts:540]: Can I check if indexing with the empty string is needed for all occurrences of `hasResponses` now?

**Review by teikjun — COMMENTED** [review]
Regarding the cause of the E2E error in `InstructorCourseEnrollPageE2ETest`:
- In `instructor-course-enroll-page.component.ts` file, `getCourseEnrollPageData` method, a modal is displayed when the course has responses 
image
- From the E2E test output, selenium is unable to click an element because  "`<div class="ng-star-inserted">` obscures it". Most likely, this div refers to the modal mentioned above. (Rationale: `ng-star-inserted` is added in some cases when `ngIf` directive is used, and the `ngIf` directive is also used for the modal). 
- So, I'm guessing that there is something wrong with the `hasResponses` field.

**Comment by moziliar** [comment]
@teikjun Ah thanks a lot for pinpointing the cause of the problem! Will look into it.

In a sense, I don't feel so good about this design, but given the little responsibility of `HasResponsesData`, I'm just make doing with the current change rn; any recommendations are appreciated!

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts (hunk 3 above)
Comment (moziliar) [reply src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts:540]: Yes; in fact, `''` should be an enum that refers to the placeholder key index for all general purpose `hasResponses` query. I guess it is currently a magic string for now which I will have to replace.

File: src/main/java/teammates/ui/webapi/GetHasResponsesAction.java (hunk 2 above)
Comment (moziliar) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:116]: Will it be too verbose in one line? Separating it this way could still work at no additional cost and improved readability.

Comment (daongochieu2810) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:116]: hmm yes I agree, lets leave it this way

File: src/main/java/teammates/ui/webapi/GetHasResponsesAction.java [hunk 4]
```diff
@@ -55,11 +59,25 @@ void checkSpecificAccessControl() {
...
+                if (feedbackSessions.isEmpty()) {
+                    // Course has no sessions and therefore no response; access to responses is safe for all.
+                    return;
+                }
 
-            gateKeeper.verifyAccessible(
-                    logic.getStudentForGoogleId(courseId, userInfo.getId()),
-                    getNonNullFeedbackSession(feedbackSessionName, courseId));
+                // Verify that all sessions are accessible to the user.
+                for (FeedbackSessionAttributes feedbackSession : feedbackSessions) {
```
Comment (daongochieu2810) [inline src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:71]: I think we should just ignore unaccessible sessions, the rest can be shown to users

Comment (moziliar) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:71]: Interesting take here: in the access control, we can't do anything but pouring all sessions into access verification; what you are proposing is ditching access verification altogether. Might need to learn about the access control within a course across sessions.

Comment (moziliar) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:71]: Taking a look at `gateKeeper.verifyAccessible`, it appears that a session is inaccessible to a student if the student is not in the course of the session, or the session is not visible.

The former doesn't matter as when we query sessions by course id, if any session is inaccessible to the student by this criterion, all the rest of the sessions shouldn't be accessed anyways; hence, we only need to take care of the latter by adding lines to skip the invisible sessions altogether.

Comment (daongochieu2810) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:71]: I see. Is it possible to extend or modify `gateKeeper.verifyAccessible` to have more fine-grained verification in this case?

Comment (moziliar) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:71]: Can clarify what do you mean by having a fine-grained verification?

Comment (daongochieu2810) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:71]: Right now we are throwing `UnauthorizedAccessException` for both cases of student not being in the course and the session not being visible. I'm proposing a separation of the 2 cases i.e having `NotInCourseException` and `NotVisibleSessionException` so we can return visible sessions and ignore sessions that cause the `NotVisibleException`

Comment (moziliar) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:71]: Was thinking of the same thing and it might be doable; regardless, `gateKeeper.verifyAccesible` is still more coupled with `hasResponseAction` in that either `hasResponseAction` has to know about the access control logic behind `verifyAccessible` or `hasResponseAction` has to know about all the exceptions `verifyAccessible` will throw and treat them accordingly.

**Review by daongochieu2810 — APPROVED** [review]
lgtm

**Review by teikjun — APPROVED** [review]

**Review by hcwong — APPROVED** [review]
some minor comments

File: src/main/java/teammates/ui/output/HasResponsesData.java [hunk 5]
```diff
@@ -1,19 +1,43 @@
...
  * The API output format to represent if there are responses.
  */
 public class HasResponsesData extends ApiOutput {
-    private final boolean hasResponses;
+    /**
+     * Placeholder key for single entry response data.
+     */
+    public static final String KEY_FOR_SINGLE_ENTRY = "";
 
+    private final Map<String, Boolean> hasResponses;
```
[the remark this answers is not in the document]
Comment (moziliar) [reply src/main/java/teammates/ui/output/HasResponsesData.java:15]: Yup, I agree with you here; else the map seems confusing. Though then again, this map doubles as single boolean with placeholder key, might need to iterate again.

File: src/main/java/teammates/storage/api/FeedbackQuestionsDb.java [hunk 6]
```diff
@@ -68,6 +68,18 @@ public FeedbackQuestionAttributes getFeedbackQuestion(
         return makeAttributes(getFeedbackQuestionEntitiesForGiverType(feedbackSessionName, courseId, giverType));
     }
 
+    /**
+     * Gets all feedback questions of a session that has certain giver type.
```
[the remark this answers is not in the document]
Comment (moziliar) [reply src/main/java/teammates/storage/api/FeedbackQuestionsDb.java]: Yep... let me fix it up.

**Review by hcwong — APPROVED** [review]

File: src/web/app/pages-student/student-home-page/student-home-page.component.ts [hunk 7]
```diff
@@ -94,22 +94,24 @@ export class StudentHomePageComponent implements OnInit {
...
-                  const isSubmitted: boolean = hasRes.hasResponses;
+            this.feedbackSessionsService.hasStudentResponseForAllFeedbackSessionsInCourse(course.courseId)
+              .subscribe((hasRes: HasResponses) => {
+                for (const fs of sortedFss) {
+                  const isOpened: boolean = fs.submissionStatus === FeedbackSessionSubmissionStatus.OPEN;
+                  const isWaitingToOpen: boolean =
+                    fs.submissionStatus === FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN;
+                  const isPublished: boolean = fs.publishStatus === FeedbackSessionPublishStatus.PUBLISHED;
+                  const isSubmitted: boolean = hasRes.hasResponsesBySession[fs.feedbackSessionName];
+                  if (isSubmitted === undefined) {
```
Comment (rrtheonlyone) [inline src/web/app/pages-student/student-home-page/student-home-page.component.ts:105]: on the topic of implementation, when would this be `undefined`? I would expect all the sessions I see in the table to have a key here.

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Nice job implementing this! I think it works @moziliar 

However, I feel that something isn't quite right. The `hasResponses` change you have implemented works well for the student home page but it makes no sense for the other use cases. If you think about it an API endpoint like `hasResponses` shouldn't be returning a map in the first place.

e.g. the empty string you have used as a key for single value returns feels weird and hard to follow when taken out of context. Imagine you are simply using the API as a black box - that result would certainly take you by surprise.  

Another idea worth considering is:
- make `hasResponses` a HEAD endpoint, based off status code, we infer the result 
- update `GetFeedbackSessionsAction` to return an extra field called `hasResponses` to indicate whether a session has responses; this field is only added when a student calls this endpoint (it looks like a student only uses this endpoint on the home page which is the primary page we are targeting)  

Let me know your thoughts on this! Regardless, your change is still an improvement and I think we can proceed otherwise too.

**Comment by moziliar** [comment]
@rrtheonlyone thanks for the thorough review. I do agree with your sentiment here. My personal take about this change is that now `hasResponses` has too many responsibilities, albeit all of which sounding justifiable under "hasResponses". While the current implementation uses a placeholder key (now set to empty string) for single response use case and meaningful keys for multiple response use case feels off from a perspective of a developer new to the code base, splitting a field out for multiple use might materialise the responsibility overloading actually, i.e. the endpoint now has multiple responsibilities and the user can only infer from the fields returned.

At first, I wish to have the single entry in the returned map keyed with meaningful name; the idea was scraped as then the key could be course id or session name, which is more confusing to me. Empty key design stems from that no session would be named empty string and therefore the only valid case is for single response return, forcing the user to infer from their input what should the return be.

I was thinking of making another endpoint that takes care of multiple session check response, but it feels odd that the old `hasResponses` endpoint still remains inefficient should some places continue to use it. However, since the current design seems off too, maybe this is more viable now?

I'm not too sure how HEAD endpoint would work as I wasn't used to using it for such use case. Perhaps I need to research more about its use and get back to you on that.

All in all, I just feel like the endpoint now kind of has more responsibilities than it should hold.

File: src/web/app/pages-student/student-home-page/student-home-page.component.ts (hunk 7 above)
Comment (moziliar) [reply src/web/app/pages-student/student-home-page/student-home-page.component.ts:105]: It's merely a defensive programming pattern I put here.

**Comment by rrtheonlyone** [comment]
> All in all, I just feel like the endpoint now kind of has more responsibilities than it should hold.

That is only after your change. Before this, the responsibility of the endpoint is clear: given a course/session/question, return a boolean flag to indicate whether they are any responses. This is why it can be a HEAD endpoint since all we need is a positive return. 

This is why I am proposing to:
- retain existing behaviour of the endpoint (and optimize if needbe)
- modify `GetFeedbackSessionsAction` to return more information so the original student home page issue is solved

**Comment by wkurniawan07** [comment]
> it can be a HEAD endpoint

HEAD request is semantically a GET request minus its response body. Unless you're only reading the status code or content length or other parts of the HTTP response (except its body, obviously), using HEAD will not work.

**Comment by rrtheonlyone** [comment]
Yup I believe we can only read status code if we just have a boolean response @wkurniawan07 - although not a big deal this. That's a separate discussion form the main issue.

**Comment by moziliar** [comment]
@rrtheonlyone so here's my understanding of the HEAD proposal:
we use HEAD endpoint for single entry `hasResponses` query and regard code 200 (or anything more appropriate) as `true`, and otherwise `false`; then for the student page use, we use GET method instead.
This could hide away the `map` in the student page multi-session return format from all other uses and does not necessarily create map entry which might be confusing.

About the proposal:

> retain existing behaviour of the endpoint (and optimize if needbe)

retain the behaviour of returning a single boolean and optimize the question fetching part still (use key only query instead)

> modify GetFeedbackSessionsAction to return more information so the original student home page issue is solved

make the `map` nullable and only return it when student home page queries it with course_id and without session id (how we determine whether to return all sessions' `hasResponses` info right now)

Am I right about the above? Personally, I feel like the HEAD way might be useful too, though the code might need to be carefully chosen (not so well versed in this). Perhaps the latter (nullable map return) is more straight forward even to new contributors?

**Comment by rrtheonlyone** [comment]
> make the map nullable and only return it when student home page queries it with course_id and without session id (how we determine whether to return all sessions' hasResponses info right now)

This is the better way forward. Let's go with this @moziliar plus leave your optimisations for `hasReponses` but still return a boolean.

The HEAD vs GET method is not something I have ever seen although it is an interesting suggestion - any APIs out there that have this?

**Comment by moziliar** [comment]
@rrtheonlyone changes have been made to support a nullable map which is read only by student home page which doesn't send session id. The original behaviour of the `hasResponsesData` is therefore retained in other places.

File: src/main/java/teammates/ui/webapi/GetHasResponsesAction.java [hunk 8]
```diff
@@ -55,11 +59,29 @@ void checkSpecificAccessControl() {
         } else {
             //An student can check whether he has submitted responses for a feedback session in his course.
             String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
-            String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+            String feedbackSessionName = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:62]: There is too much nesting in `checkSpecificAccessControl` - let's clean this up a little to make it more readable. 

Do:
```
if (check) {

    return;
}

//rest of code
```

Instead of:
```
if (check) {

} else {
  //rest of code
}
```

File: src/main/java/teammates/ui/webapi/GetHasResponsesAction.java [hunk 9]
```diff
@@ -87,17 +109,33 @@ JsonResult execute() {
             return new JsonResult(new HasResponsesData(hasResponses));
         } else {
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:110]: this else is unnecessary - we already `return` above. This is not your change but please help to clean this up as well (since we are on this page)

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Sorry for the delay in this. Really Nice work 💯 

I have put in some minor code quality improvement suggestions below.

Before we merge in, please also:
- Update `GetHasResponsesActionTest` to include the new cases we have introduced
- Update `student-home-page.component.spec.ts` to have a snapshot test (to ensure table get's populated correct based on API return)

File: src/web/services/feedback-sessions.service.ts [hunk 10]
```diff
@@ -202,6 +202,18 @@ export class FeedbackSessionsService {
     return this.httpRequestService.get(ResourceEndpoints.HAS_RESPONSES, paramMap);
   }
 
+  /**
+   * Checks if there is response of a student for an array of feedback sessions (request sent by student).
+   */
+  hasStudentResponseForAllFeedbackSessionsInCourse(courseId: string): Observable<HasResponses> {
```
[the remark this answers is not in the document]
Comment (moziliar) [reply src/web/services/feedback-sessions.service.ts:208]: At this point, the function name only serves to distinguish the purpose of the API call. This is done purely out of a better semantics.

File: src/main/java/teammates/ui/webapi/GetHasResponsesAction.java (hunk 8 above)
Comment (moziliar) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:62]: I will just remove any redundant `else` clause that is immediately succeeded by function return

File: src/main/java/teammates/ui/webapi/GetHasResponsesAction.java (hunk 9 above)
Comment (moziliar) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:110]: Got it! I do agree with some cleanup here!

Comment (moziliar) [reply src/main/java/teammates/ui/webapi/GetHasResponsesAction.java:110]: After some tinkering, removal of the `else` clause results in an un-bypassable lint failure from the nested `if` on top from pmdMain. Have to extract the instructor logic away into an individual method to avoid this, which should be more readable in a sense.

File: src/web/app/pages-student/student-home-page/student-home-page.component.ts [hunk 11]
```diff
@@ -94,22 +94,27 @@ export class StudentHomePageComponent implements OnInit {
...
-                .subscribe((hasRes: HasResponses) => {
-                  const isSubmitted: boolean = hasRes.hasResponses;
+            this.feedbackSessionsService.hasStudentResponseForAllFeedbackSessionsInCourse(course.courseId)
+              .subscribe((hasRes: HasResponses) => {
+                for (const fs of sortedFss) {
+                  const isOpened: boolean = fs.submissionStatus === FeedbackSessionSubmissionStatus.OPEN;
+                  const isWaitingToOpen: boolean =
+                    fs.submissionStatus === FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN;
+                  const isPublished: boolean = fs.publishStatus === FeedbackSessionPublishStatus.PUBLISHED;
+                  if (hasRes.hasResponsesBySession === undefined) {
```
[the remark this answers is not in the document]
Comment (moziliar) [reply src/web/app/pages-student/student-home-page/student-home-page.component.ts:104]: We are using optional `hasResponsesBySession` here to determine if the call is for a single session or multiple sessions. The `undefined` check here is required by the linter.

[the remark this answers is not in the document]
Comment (rrtheonlyone) [reply src/web/app/pages-student/student-home-page/student-home-page.component.ts:104]: I think we can pre-check outside the loop - let me try it out and see if that works out. I will check in again with you on this.

File: src/web/services/feedback-sessions.service.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (rrtheonlyone) [reply src/web/services/feedback-sessions.service.ts:208]: Yes that is fair - no issues

**Review by rrtheonlyone — APPROVED** [review]
Thanks @moziliar we are good here 💯 

I have just did some minor touch-ups on my end so we can move forward:
- Did a pre-check that API output is valid on FE and print error to user (no need to `continue` in for loop or keep checking the same map)
- Added both FE and BE tests to capture the new logic

This way we ensure the output is exactly the same before the change + a performance increase.
